vendor/diff-stdlib-format.patch

```diff
@@ -0,0 +1,222 @@
+--- ocaml-4.13-upstream/format.ml
++++ ocamlformat_support/format.ml
+@@@@
+   | Pp_tbegin of tbox          (* beginning of a tabulation box *)
+   | Pp_tend                    (* end of a tabulation box *)
+   | Pp_newline                 (* to force a newline inside a box *)
+   | Pp_if_newline              (* to do something only if this very
+                                   line has been broken *)
++  | Pp_string_if_newline of string
++                               (* print a string only if this very
++                                  line has been broken *)
++  | Pp_or_newline of int * int * string * string
++                               (* print a break and the first string if this
++                                  very line has not been broken, otherwise
++                                  print the second string *)
++  | Pp_fits_or_breaks of int * string * int * int * string
++                               (* print a string if the enclosing box fits,
++                                  otherwise print a break and a string *)
+   | Pp_open_tag of stag         (* opening a tag name *)
+   | Pp_close_tag               (* closing the most recently open tag *)
+ 
+ and stag = ..
+ 
+@@@@
+   format_string state before;
+   pp_output_newline state;
+   state.pp_is_new_line <- true;
+   let indent = state.pp_margin - width + offset in
+   (* Don't indent more than pp_max_indent. *)
+-  let real_indent = Int.min state.pp_max_indent indent in
++  let real_indent = min state.pp_max_indent indent in
+   state.pp_current_indent <- real_indent;
+   state.pp_space_left <- state.pp_margin - state.pp_current_indent;
+   pp_output_indent state state.pp_current_indent;
+   format_string state after
+ 
+@@@@
+   The main pretty printing functions.
+ 
+ *)
+ 
+ (* Formatting a token with a given size. *)
+-let format_pp_token state size = function
++let rec format_pp_token state size = function
+ 
+   | Pp_text s ->
+     format_pp_text state size s
+ 
+   | Pp_begin (off, ty) ->
+@@@@
+ 
+   | Pp_if_newline ->
+     if state.pp_current_indent != state.pp_margin - state.pp_space_left
+     then pp_skip_token state
+ 
++  | Pp_string_if_newline s ->
++    if state.pp_is_new_line
++    then format_string state s
++
+   | Pp_break { fits; breaks } ->
+     let before, off, _ = breaks in
+     begin match Stack.top_opt state.pp_format_stack with
+     | None -> () (* No open box. *)
+     | Some { box_type; width } ->
+@@@@
+       | Pp_vbox -> break_new_line state breaks width
+       | Pp_hbox -> break_same_line state fits
+       end
+     end
+ 
++  | Pp_or_newline (n, off, fits, breaks) ->
++    if state.pp_is_new_line
++    then format_string state breaks
++    else
++      let fits = ("", n, fits) and breaks = ("", off, breaks) in
++      format_pp_token state size (Pp_break {fits; breaks})
```

Comment:
Maybe it is better as a diff-reduction approach, but I would add a comment that when upstreaming this, it would still be better to make the `Pp_break` case into a function and also call it here instead of making `format_pp_token` recursive.

Reply:
I will revert this change to keep the function instead, that will be clearer and avoid missing it when upstreaming